### ISSUE TITLE
hooks / useQuery optimizations

### ISSUE THREAD

**Issue by tantaman** [vlcn-io/cr-sqlite#78]
# Prepare, not exec

the majority of time spent in simple queries is spent compiling the statement given that statement compilation requires:
1. parsing the statement
2. calling all tables involved to compute index costs
3. creating a query plan
4. generating the bytecode for that plan

calling `exec` has to re-do compilation. We can save quite a bit of time and work by preparing the statement in `useQuery`.

# vlcn value

Make use of PSD and/or STM to commit all changes to memory at once and only incur a single react re-render.
This is required given the SQLite query interface is async which doesn't play well with react's auto-batching given event loop control keeps passing in and out of react and queries.

To do this:
1. The `rx` hook would start a STM or PSD transaction when invoking listeners
2. A tx complete we set all the state values

```
rx.notify {
  STM.txAsync(() => 
    // this step is responsible for re-running queries
    Promise.all(listeners.map(l => l());
    // now we can release the live results once all reactive queries completed?
  );
}
```

^-- drawback here is you end up waiting on the slowest query.
^-- Upside is things change transactionally rather than seeing partial state.

# Dependent queries in the same component

```
function Comp() {
  const r1 = useQuery(`select f from s`);
  const r2 = useQuery(`select b from d where x = ${r1}`);
}
```

Do we need to do anything here? You could hit cases where r1 & r2 run reactively at t0 then r2 runs again at t1 via being modified by r1's return value, throwing away the reactive r2 at t0 result.

In other words, we can remove reactive computations in certain situations. This is out of scope for now.

**Comment by tantaman** [comment]
see https://github.com/Effect-TS/examples/tree/9940cf90e04a4acd8c57aee4b442b1a847778ff1/react-heavy-work for other prior art

**Comment by tantaman** [comment]
Mostly resolved by: https://github.com/vlcn-io/cr-sqlite/pull/133

**Comment by tantaman** [comment]
We can also do something simple to start like create an adaptor for TanstackQuery.

**Comment by tantaman** [comment]
Updated further to run all calls to `useQuery` within the same read transaction.

https://github.com/vlcn-io/cr-sqlite/pull/159
